ansible_rulebook/terminal.py

```diff
@@ -0,0 +1,191 @@
+#  Copyright 2023 Red Hat, Inc.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import datetime
+import logging
+import os
+import pprint
+import re
+import threading
+import typing
+
+logger = logging.getLogger(__name__)
+
+
+###############################################################################
+class _Singleton(type):
+    def __init__(cls, name, bases, dct, **kwargs) -> None:
+        super().__init__(name, bases, dct, **kwargs)
+        cls.__singleton = None
+        cls.__lock = threading.RLock()
```

Comment:
What is the need for a lock?

I think this might be a potential issue. Ansible-rulebook is essentially an asyncio-based application. Asyncio is designed for cooperative multitasking and doesn't mix well with traditional thread synchronization primitives like locks. When you use a blocking lock in an asyncio application, you risk blocking the entire event loop, which can lead to performance issues or deadlocks.

Instead, I propose considering `asyncio.lock()` https://docs.python.org/3/library/asyncio-sync.html#asyncio.Event if locking is really necessary. Alternatively, we could reconsider the whole singleton pattern in favor of dependency injection or using shared instances. These alternatives fit more naturally with asyncio's paradigms.

Reply:
The lock is only present to synchronize the creation of the singleton and is held only long enough to do so.  Once the singleton is created the lock is never again utilized.  It should not be an issue.
But using asyncio.lock() should also suffice.

Reply:
Using asyncio, with its imposed requirements on methods, doesn't play well with the metaclassing (object hierarchy creation) and use of the descendant class.  Given the limited nature of the lock use I don't see a need for changes here.